# Superstore — Python Analysis (BQ6–BQ10 + Advanced)
**Dataset:** Sample_Superstore.csv


| File | Phạm vi |
|------|---------|
| `Superstore_SQL.sql` | BQ1–BQ5 |
| `Superstore_Python.ipynb` | BQ6–BQ10 + Advanced |


## 0. Setup & Load Data

In [ ]:
import pandas as pd
import numpy as np
from scipy import stats

df = pd.read_csv('/content/Superstore.csv', encoding='latin-1')
df['Order Date'] = pd.to_datetime(df['Order Date'])
df['Ship Date']  = pd.to_datetime(df['Ship Date'])
df['Year']       = df['Order Date'].dt.year
df['Month']      = df['Order Date'].dt.month

num_rows = len(df)
num_cols = df.shape[1]
min_order_date = df['Order Date'].min().strftime('%d/%m/%Y')
max_order_date = df['Order Date'].max().strftime('%d/%m/%Y')
unique_orders = df['Order ID'].nunique()
unique_customers = df['Customer ID'].nunique()
total_sales = df['Sales'].sum()
total_profit = df['Profit'].sum()
missing_values = df.isnull().sum().sum()

print(f"Số dòng: {num_rows:,}")
print(f"Số cột: {num_cols}")
print(f"Giai đoạn: {min_order_date} – {max_order_date}")
print(f"Đơn hàng unique: {unique_orders:,}")
print(f"Khách hàng unique: {unique_customers:,}")
print(f"Tổng doanh thu: ${total_sales:,.0f}")
print(f"Tổng lợi nhuận: ${total_profit:,.0f}")
print(f"Missing values: {missing_values}")

Số dòng: 9,994
Số cột: 23
Giai đoạn: 03/01/2014 – 30/12/2017
Đơn hàng unique: 5,009
Khách hàng unique: 793
Tổng doanh thu: $2,297,201
Tổng lợi nhuận: $286,397
Missing values: 0


---
## BQ6 — % đóng góp doanh thu Sub-Category trong Category

**Câu hỏi:** Mỗi Sub-Category chiếm bao nhiêu % doanh thu của Category chứa nó?


In [ ]:
# BQ6
bq6 = (df.groupby(['Category', 'Sub-Category'])
         .agg(sales=('Sales','sum'), profit=('Profit','sum'))
         .reset_index())

# transform('sum') = SUM() OVER (PARTITION BY Category)
bq6['cat_total']  = bq6.groupby('Category')['sales'].transform('sum')
bq6['pct_of_cat'] = (bq6['sales'] / bq6['cat_total'] * 100).round(1)
bq6['margin_pct'] = (bq6['profit'] / bq6['sales'] * 100).round(1)
bq6 = bq6.sort_values(['Category', 'pct_of_cat'], ascending=[True, False])

print(bq6[['Category','Sub-Category','sales','profit','pct_of_cat','margin_pct']]
      .rename(columns={'sales':'Sales','profit':'Profit',
                       'pct_of_cat':'% of Category','margin_pct':'Margin%'})
      .to_string(index=False))


       Category Sub-Category       Sales      Profit  % of Category  Margin%
      Furniture       Chairs 328449.1030  26590.1663           44.3      8.1
      Furniture       Tables 206965.5320 -17725.4811           27.9     -8.6
      Furniture    Bookcases 114879.9963  -3472.5560           15.5     -3.0
      Furniture  Furnishings  91705.1640  13059.1436           12.4     14.2
Office Supplies      Storage 223843.6080  21278.8264           31.1      9.5
Office Supplies      Binders 203412.7330  30221.7633           28.3     14.9
Office Supplies   Appliances 107532.1610  18138.0054           15.0     16.9
Office Supplies        Paper  78479.2060  34053.5693           10.9     43.4
Office Supplies     Supplies  46673.5380  -1189.0995            6.5     -2.5
Office Supplies          Art  27118.7920   6527.7870            3.8     24.1
Office Supplies    Envelopes  16476.4020   6964.1767            2.3     42.3
Office Supplies       Labels  12486.3120   5546.2540            1.7     44.4

**Kết quả:**
- Chairs chiếm **44.3%** Furniture sales — lớn nhất nhóm
- Tables chiếm **27.9%** Furniture sales nhưng margin **-8.6%** → lỗ ròng -$17,725
- Phones chiếm **39.5%** Technology sales
- Copiers chỉ **17.9%** Technology sales nhưng margin **37.2%** → sinh lời nhất

---
## BQ7 — Doanh thu từng tháng 2017 vs 2016

**Câu hỏi:** Mỗi tháng năm 2017 tăng hay giảm so với cùng kỳ 2016?



In [ ]:
# BQ7 — pivot year thành cột thay cho self-join SQL
monthly = (df[df['Year'].isin([2016, 2017])]
           .groupby(['Year', 'Month'])['Sales']
           .sum()
           .unstack(level=0)   # Year → column: đây là lý do Python gọn hơn SQL
           .round(2))
monthly.columns = ['sales_2016', 'sales_2017']
monthly['sales_diff'] = (monthly['sales_2017'] - monthly['sales_2016']).round(2)
monthly['growth_pct'] = (monthly['sales_diff'] / monthly['sales_2016'] * 100).round(1)

MONTHS = {1:'Jan',2:'Feb',3:'Mar',4:'Apr',5:'May',6:'Jun',
          7:'Jul',8:'Aug',9:'Sep',10:'Oct',11:'Nov',12:'Dec'}
monthly.index = monthly.index.map(MONTHS)

print(monthly.to_string())
print(f"\nTổng 2017: ${monthly.sales_2017.sum():,.0f}")
print(f"Tổng 2016: ${monthly.sales_2016.sum():,.0f}")
print(f"Net diff : ${monthly.sales_diff.sum():,.0f}  ({monthly.sales_diff.sum()/monthly.sales_2016.sum()*100:.1f}%)")

       sales_2016  sales_2017  sales_diff  growth_pct
Month                                                
Jan      18542.49    43971.37    25428.88       137.1
Feb      22978.82    20301.13    -2677.69       -11.7
Mar      51715.88    58872.35     7156.47        13.8
Apr      38750.04    36521.54    -2228.50        -5.8
May      56987.73    44261.11   -12726.62       -22.3
Jun      40344.53    52981.73    12637.20        31.3
Jul      39261.96    45264.42     6002.46        15.3
Aug      31115.37    63120.89    32005.52       102.9
Sep      73410.02    87866.65    14456.63        19.7
Oct      59687.74    77776.92    18089.18        30.3
Nov      79411.97   118447.82    39035.85        49.2
Dec      96999.04    83829.32   -13169.72       -13.6

Tổng 2017: $733,215
Tổng 2016: $609,206
Net diff : $124,010  (20.4%)


**Kết quả thực:**
- Tháng 1: **+137.1%** — tăng mạnh nhất
- Tháng 8: **+102.9%** — tháng thứ 2 tăng hơn 100%
- Tháng 5: **-22.3%** — giảm mạnh nhất
- Tháng 12: **-13.6%** — cuối năm sụt (dù Q4 thường cao)
- Tổng 2017: $733,215 vs 2016: $609,206 → **+$124,009 (+20.4%)**

---
## BQ8 — Khách hàng Loyal (mua ≥ 3 năm khác nhau)

**Câu hỏi:** Ai là khách hàng trung thành? VIP top sales nào đang thực sự sinh lời?


In [ ]:
# BQ8
cust = (df.groupby(['Customer Name', 'Segment'])
          .agg(years_active=('Year',    'nunique'),
               total_orders=('Order ID','nunique'),
               total_sales =('Sales',   'sum'),
               total_profit=('Profit',  'sum'))
          .reset_index())
cust['margin_pct']   = (cust['total_profit'] / cust['total_sales'] * 100).round(1)
cust['avg_discount'] = (df.groupby('Customer Name')['Discount'].mean() * 100).round(1).values

loyal = cust[cust['years_active'] >= 3].sort_values('total_sales', ascending=False).copy()
loyal['vip_flag'] = loyal['total_profit'].apply(
    lambda p: 'Net Loss VIP'  if p < 0
         else 'Low Margin VIP' if p/loyal.loc[loyal.total_profit==p,'total_sales'].values[0] < 0.05
         else 'Profitable')
# Đơn giản hơn với apply trực tiếp:
loyal['vip_flag'] = loyal.apply(
    lambda r: 'Net Loss VIP'   if r.total_profit < 0
         else 'Low Margin VIP' if r.margin_pct   < 5
         else 'Profitable', axis=1)

print(f"Total customers : {len(cust)}")
print(f"Loyal (≥3 yrs)  : {len(loyal)} ({len(loyal)/len(cust):.1%})")
print(f"Net Loss VIPs   : {(loyal.vip_flag=='Net Loss VIP').sum()}")
print()
print("Top 15 Loyal by Sales:")
cols = ['Customer Name','Segment','years_active','total_orders',
        'total_sales','total_profit','margin_pct','vip_flag']
print(loyal.head(15)[cols].to_string(index=False))


Total customers : 793
Loyal (≥3 yrs)  : 628 (79.2%)
Net Loss VIPs   : 117

Top 15 Loyal by Sales:
 Customer Name     Segment  years_active  total_orders  total_sales  total_profit  margin_pct     vip_flag
   Sean Miller Home Office             3             5    25043.050    -1980.7393        -7.9 Net Loss VIP
  Tamara Chand   Corporate             3             5    19052.218     8981.3239        47.1   Profitable
  Tom Ashbrook Home Office             3             4    14595.620     4703.7883        32.2   Profitable
 Adrian Barton    Consumer             4            10    14473.571     5444.8055        37.6   Profitable
  Ken Lonsdale    Consumer             4            12    14175.229      806.8550         5.7   Profitable
  Sanjit Chand    Consumer             4             9    14142.334     5757.4119        40.7   Profitable
  Hunter Lopez    Consumer             3             6    12873.298     5622.4292        43.7   Profitable
  Sanjit Engle    Consumer             4      

**Kết quả thực:**
- **628 / 793** khách hàng loyal = **79.2%** — tỉ lệ cao
- Sean Miller: $25,043 sales nhưng `Lỗ-$1,981` → "expensive VIP

- Tamara Chand: $19,052 sales, `Lãi $8,981` (margin 47.1%) → VIP thực sự
- 117 / 628 loyal customers đang có profit âm

---
## BQ9 — Sub-Category sinh lời cao nhất mỗi Segment

**Câu hỏi:** Mỗi Customer Segment nên tập trung đẩy Sub-Category nào?

In [ ]:
# BQ9
bq9 = (df.groupby(['Segment', 'Sub-Category', 'Category'])
         .agg(total_sales=('Sales','sum'), total_profit=('Profit','sum'))
         .reset_index())
bq9['margin_pct']  = (bq9['total_profit'] / bq9['total_sales'] * 100).round(1)
bq9['rank_in_seg'] = (bq9.groupby('Segment')['total_profit']
                         .rank(ascending=False, method='min')
                         .astype(int))
bq9 = bq9.sort_values(['Segment', 'rank_in_seg'])

print("=== Top 3 Sub-Categories per Segment ===")
print(bq9[bq9['rank_in_seg'] <= 3]
      [['Segment','rank_in_seg','Sub-Category','total_profit','margin_pct']]
      .to_string(index=False))

print("\n=== Loss Sub-Categories per Segment ===")
print(bq9[bq9['total_profit'] < 0]
      [['Segment','Sub-Category','total_profit','margin_pct']]
      .sort_values(['Segment','total_profit'])
      .to_string(index=False))


=== Top 3 Sub-Categories per Segment ===
    Segment  rank_in_seg Sub-Category  total_profit  margin_pct
   Consumer            1      Copiers    24083.7106        34.5
   Consumer            2       Phones    23837.1147        14.0
   Consumer            3  Accessories    20735.9225        23.8
  Corporate            1      Copiers    18990.2789        40.6
  Corporate            2  Accessories    12707.4805        26.4
  Corporate            3       Phones    11766.2196        12.9
Home Office            1      Copiers    12543.8354        38.2
Home Office            2       Phones     8912.3963        12.9
Home Office            3  Accessories     8493.2327        26.5

=== Loss Sub-Categories per Segment ===
    Segment Sub-Category  total_profit  margin_pct
   Consumer       Tables    -9728.0378        -9.7
   Consumer    Bookcases    -4435.6382        -6.5
   Consumer     Supplies    -1657.5513        -6.4
  Corporate       Tables    -4906.4986        -6.9
Home Office       Table

**Kết quả thực:**
- **Copiers** là #1 profit ở **cả 3 Segments** (margin 34–41%)
- **Home Office + Office Supplies** = margin **20.8%** — cao nhất toàn matrix
- Tables và Supplies lỗ ở tất cả Segments


---
## BQ10 — Phân loại đơn hàng High / Medium / Low theo Region

**Câu hỏi:** Mỗi Region có phân bổ đơn hàng theo giá trị như thế nào?


In [ ]:
# BQ10
order_val = (df.groupby(['Order ID', 'Region'])['Sales']
               .sum().reset_index(name='order_sales'))

order_val['value_tier'] = pd.cut(
    order_val['order_sales'],
    bins=[-1, 500, 1000, 1e9],
    labels=['Low (<$500)', 'Medium ($500–1K)', 'High (>$1K)'])

bq10 = (order_val.groupby(['Region', 'value_tier'], observed=True)
                 .agg(total_orders=('Order ID','count'),
                      total_sales =('order_sales','sum'))
                 .reset_index())
bq10['pct_orders'] = (bq10['total_orders']
                      / bq10.groupby('Region')['total_orders'].transform('sum') * 100).round(1)
bq10['total_sales'] = bq10['total_sales'].round(2)

print(bq10.to_string(index=False))

# Pivot để dễ đọc
print("\n=== Pivot — Order Count ===")
print(bq10.pivot(index='Region', columns='value_tier', values='total_orders')
          .to_string())
print("\n=== Pivot — % of Region Orders ===")
print(bq10.pivot(index='Region', columns='value_tier', values='pct_orders')
          .to_string())


 Region       value_tier  total_orders  total_sales  pct_orders
Central      Low (<$500)           892    110383.69        75.9
Central Medium ($500–1K)           154    107564.80        13.1
Central      High (>$1K)           129    283291.40        11.0
   East      Low (<$500)          1034    135945.59        73.8
   East Medium ($500–1K)           193    135161.89        13.8
   East      High (>$1K)           174    407673.77        12.4
  South      Low (<$500)           603     75245.02        73.4
  South Medium ($500–1K)           117     83836.02        14.2
  South      High (>$1K)           102    232640.86        12.4
   West      Low (<$500)          1206    167487.08        74.9
   West Medium ($500–1K)           190    135356.89        11.8
   West      High (>$1K)           215    422613.85        13.3

=== Pivot — Order Count ===
value_tier  Low (<$500)  Medium ($500–1K)  High (>$1K)
Region                                                
Central             892      

**Kết quả thực:**
- Tất cả 4 Regions: Low (<$500) chiếm ~74% số đơn — majority
- West có nhiều High-value orders nhất (215 đơn)
- Central: tỉ lệ High orders 11% nhưng margin vẫn thấp nhất (7.9%)
  → vấn đề không nằm ở order size mà ở discount policy

---
## Advanced — RFM Segmentation


In [ ]:
# RFM Segmentation
SNAPSHOT = pd.Timestamp('2018-01-01')

rfm = (df.groupby('Customer Name')
         .agg(recency   =('Order Date', lambda x: (SNAPSHOT - x.max()).days),
              frequency  =('Order ID',   'nunique'),
              monetary   =('Sales',      'sum'))
         .round(2))

# pd.qcut() chia đều 5 nhóm theo phân vị — thay NTILE(5) + RANKX trong DAX
rfm['R'] = pd.qcut(rfm['recency'],              5, labels=[5,4,3,2,1]).astype(int)
rfm['F'] = pd.qcut(rfm['frequency'].rank(method='first'), 5, labels=[1,2,3,4,5]).astype(int)
rfm['M'] = pd.qcut(rfm['monetary'],             5, labels=[1,2,3,4,5]).astype(int)
rfm['RFM_Score'] = rfm['R'] + rfm['F'] + rfm['M']

def rfm_label(s):
    if s >= 13: return 'Champions'
    if s >= 10: return 'Loyal'
    if s >= 7:  return 'Potential'
    if s >= 5:  return 'At Risk'
    return 'Lost'

rfm['Segment'] = rfm['RFM_Score'].apply(rfm_label)
rfm = rfm.reset_index()

# Summary
seg_summary = (rfm.groupby('Segment')
                  .agg(customers =('Customer Name','count'),
                       avg_recency  =('recency','mean'),
                       avg_frequency=('frequency','mean'),
                       avg_monetary =('monetary','mean'))
                  .round(1))
SEG_ORDER = ['Champions','Loyal','Potential','At Risk','Lost']
print("=== RFM Segment Summary ===")
print(seg_summary.reindex(SEG_ORDER).to_string())

print("\n=== Full RFM Table (top 20 by score) ===")
print(rfm.nlargest(20,'RFM_Score')
         [['Customer Name','recency','frequency','monetary','R','F','M','RFM_Score','Segment']]
         .to_string(index=False))


=== RFM Segment Summary ===
           customers  avg_recency  avg_frequency  avg_monetary
Segment                                                       
Champions        124         29.1            9.4        5221.3
Loyal            255         76.1            7.4        3862.7
Potential        217        138.0            5.6        2159.2
At Risk          118        229.5            4.2        1239.4
Lost              79        480.4            3.0         632.9

=== Full RFM Table (top 20 by score) ===
    Customer Name  recency  frequency  monetary  R  F  M  RFM_Score   Segment
       Alan Hwang        8          9   4805.34  5  5  5         15 Champions
   Arianne Irving       15         10   4375.79  5  5  5         15 Champions
       Ben Ferrer       18         11   5907.97  5  5  5         15 Champions
  Dan Reichenbach        5          9   6528.03  5  5  5         15 Champions
        Dean Katz       12          9   4802.39  5  5  5         15 Champions
       Emily Phan    

**Kết quả thực (793 customers):**
Champions 124 · Loyal 255 · Potential 217 · At Risk 118 · Lost 79

---
## Export

In [ ]:
# Chạy cell này để export tất cả kết quả ra CSV
# → Import vào Power BI bằng Get Data > Text/CSV

# BQ6
bq6.to_csv('export_bq6_subcat_pct.csv', index=False)

# BQ7 (Đã đổi tên cột tránh lỗi trùng hàm hệ thống)
monthly.to_csv('export_bq7_yoy_monthly.csv')

# BQ8
loyal.to_csv('export_bq8_loyal_customers.csv', index=False)

# BQ9
bq9.to_csv('export_bq9_segment_subcat.csv', index=False)

# BQ10
bq10.to_csv('export_bq10_order_tier.csv', index=False)

# RFM
rfm.to_csv('export_rfm_segments.csv', index=False)


✓ Exported 7 files — import vào Power BI để build visual.
  Xem Power_BI_Guide_BQ6_BQ10.md để biết cách dựng từng chart.
